### ISSUE TITLE
opam-dune-lint suggests depending on libraries that are transitive dependencies

### ISSUE THREAD

**Issue by tmcgilchrist** [ocurrent/opam-dune-lint#57]
Perhaps related to this [#53](https://github.com/ocurrent/opam-dune-lint/issues/53) but I'm getting this report on ocaml-grpc. Which is suggesting the wrong library versions `"0"` and is asking for libraries to be listed that are coming from a dependency library.
eg grpc-examples depends on grpc-async, which is how it picks up async. It shouldn't need to depend directly on async. 
Similarly for grpc-async it depends on grpc which depends on h2, so I shouldn't need a direct dependency on it in grpc-async's opam file.

```
grpc-async.opam: changes needed:
  "h2" {>= "0"}                            [from lib/grpc-async]
  "ppx_jane" {>= "0"}                      [from lib/grpc-async]
grpc-bench.opam: OK
grpc-eio.opam: changes needed:
  "h2" {>= "0"}                            [from lib/grpc-eio]
grpc-examples.opam: changes needed:
  "async" {>= "0"}                         [from examples/greeter-async]
  "cohttp" {>= "0"}                        [from examples/etcd]
  "cohttp-lwt" {>= "0"}                    [from examples/etcd]
  "core" {>= "0"}                          [from examples/greeter-async, examples/routeguide-async/proto]
  "core_unix" {>= "0"}                     [from examples/greeter-async]
  "h2" {>= "0"}                            [from examples/etcd, examples/greeter-ssl-lwt]
  "lwt" {>= "0"}                           [from examples/etcd, examples/greeter-ssl-lwt]
  "ppx_deriving" {>= "0"}                  [from examples/etcd, examples/routeguide-async/proto, examples/routeguide/proto]
  "ppx_jane" {>= "0"}                      [from examples/greeter-async, examples/routeguide-async/proto]
  "uri" {>= "0"}                           [from examples/etcd]
grpc-lwt.opam: changes needed:
  "h2" {>= "0"}                            [from lib/grpc-lwt]
grpc.opam: changes needed:
  "bigstringaf" {>= "0"}                   [from lib/grpc]
```

Build logs at https://ocaml.ci.dev/github/dialohq/ocaml-grpc/commit/3886b3187023fbbda989239f6055ffa2d61afa29/variant/%28lint-opam%29

**Comment by moyodiallo** [comment]
This is not related to https://github.com/ocurrent/opam-dune-lint/issues/53, it is slightly different, here the suggested libraries are found in the same directory as the package which is correct.

>eg grpc-examples depends on grpc-async, which is how it picks up async. It shouldn't need to depend directly on async.
Similarly for grpc-async it depends on grpc which depends on h2, so I shouldn't need a direct dependency on it in grpc-async's opam file.

In the case of grpc-async, `h2` is coming from `lib/grpc-async/dune` not `lib/grpc/dune`:
```
...
"h2" {>= "0"}                            [from lib/grpc-async]
...
```
And in the case of grpc-examples, `async` is found in `lib/examples/greeter-async/dune` not in `lib/grpc-async/dune`:
```
...
"async" {>= "0"}                         [from examples/greeter-async]
...
```
The transitive dependency is not resolved for a  package aka public library when another lib depends on that package. Eg the dependencies of gprc.opam are not resolved when linting gprc-async.opam. All the packages(public libs) are independently resolved from each other.

The versions is correct because those libraries are not install before the lint.
```
WARNING: can't find opam package providing "h2"!
WARNING: can't find opam package providing "ppx_jane"!
```
